Objective-C/LogSink/CBLLogSinks.mm

```diff
@@ -0,0 +1,211 @@
+static CBLLogLevel _callbackLevel = kCBLLogLevelNone;
+static LogAPI _vAPI = LogAPINone;
+static BOOL isInitialized = YES;
+
+static CBLConsoleLogSink* _console = nil;
+static CBLCustomLogSink* _custom = nil;
+static CBLFileLogSink* _file = nil;
+
+// Note:
+//
+// This class is implemented with a minimum logging. The interface defines
+// console, custom, and file property as nonatomic, so the expectation
+// is not to set up the log sinks object concurrently. However, at least,
+// the lock to access console and custom is required to avoid crash from accessing
+// garbage objects as LiteCore's log callback may be called from a background thread.
+
+@implementation CBLLogSinks
+
++ (void) initialize {
+    if (self == [CBLLogSinks class]) {
+        [self init];
+        isInitialized = NO;
+    }
+}
+
+NSDictionary* domainDictionary = nil;
+
++ (void) init {
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        // Cache log domain for logging from the platforms:
+        kCBL_LogDomainDatabase  = c4log_getDomain("DB", true);
+        kCBL_LogDomainQuery  = c4log_getDomain("Query", true);
+        kCBL_LogDomainSync  = c4log_getDomain("Sync", true);
+        kCBL_LogDomainWebSocket  = c4log_getDomain("WS", true);
+        kCBL_LogDomainListener  = c4log_getDomain("Listener", true);
+        
+        // Create the default warning console log:
+        self.console = [[CBLConsoleLogSink alloc] initWithLevel: kCBLLogLevelWarning];
+    });
+}
+
++ (void) setConsole:(CBLConsoleLogSink*)console {
+    CBL_LOCK(self) {
+        [self checkLogApiVersion: LogAPINew];
+        _console = console;
+    }
+    [self updateLogLevels];
+}
+
++ (CBLConsoleLogSink*) console {
+    CBL_LOCK(self) {
+        [self checkLogApiVersion: LogAPINew];
+        return _console;
+    }
+}
+
++ (void) setCustom: (CBLCustomLogSink*) custom {
+    CBL_LOCK(self) {
+        [self checkLogApiVersion: LogAPINew];
+        _custom = custom;
+    }
+    [self updateLogLevels];
+}
+
++ (CBLCustomLogSink*) custom {
+    CBL_LOCK(self) {
+        [self checkLogApiVersion: LogAPINew];
+        return _custom;
+    }
+}
+
++ (void) setFile: (CBLFileLogSink*) file {
+    CBL_LOCK(self) {
+        [self checkLogApiVersion: LogAPINew];
+    }
+    
+    _file = file;
+    [CBLFileLogSink setup: file];
+    [self updateLogLevels];
+}
+
++ (CBLFileLogSink*) file {
+    CBL_LOCK(self) {
+        [self checkLogApiVersion: LogAPINew];
+        return _file;
+    }
+}
+
++ (void) updateLogLevels {
+    CBLConsoleLogSink* console = self.console;
+    CBLLogLevel consoleLevel = console != nil ? console.level : kCBLLogLevelNone;
+    
+    CBLCustomLogSink* custom = self.custom;
+    CBLLogLevel customLevel = custom.logSink != nil ? custom.level : kCBLLogLevelNone;
+    
+    CBLFileLogSink* file = self.file;
+    CBLLogLevel fileLevel = file != nil ? file.level : kCBLLogLevelNone;
+    
+    CBLLogLevel callbackLevel = std::min(consoleLevel, customLevel);
```

Comment:
At first I thought that the two vars were redundant... but this is really nice and readable.